File: SL.Web/src/slApiClient.ts
import {
  AppSettings,
  Arrival,
  InferredBusPosition,
  LineConfig,
  StopGroup,
  StationRouteSuggestions,
  TripCall,
  UpcomingBusArrival,
  UpcomingBusLineSummary,
  UpcomingBusQueryResult
} from "./types";

const DEFAULT_BASE_URL = "https://realtime-api.trafiklab.se/v1";
const STATIONS_CACHE_KEY = "sl_all_stations_cache_v1";
const STATIONS_CACHE_TTL_MS = 24 * 60 * 60 * 1000;

export class SLApiClient {
  private readonly apiKey: string;
  private readonly baseUrl: string;
  private static stationsMemoryCache: string[] | null = null;

  constructor(apiKey: string, baseUrl: string = DEFAULT_BASE_URL) {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl;
  }

  async getUpcomingBuses(settings: AppSettings, now: Date = new Date()): Promise<UpcomingBusQueryResult | null> {
    const allArrivals: UpcomingBusArrival[] = [];
    const lineSummaries: UpcomingBusLineSummary[] = [];

    for (const line of settings.lines) {
      const stopGroups = await this.findStopGroups(line.stopGroupSearchName);
      if (stopGroups.length === 0) continue;

      for (const stopGroup of stopGroups) {
        const arrivals = await this.getArrivals(stopGroup.id);
        const filtered = this.filterArrivals(arrivals, line, settings.timeWindowMinutes, now)
          .filter((arrival) => arrival.route.designation.trim().length > 0);

        allArrivals.push(
          ...filtered
            .map((arrival) => this.mapArrival(arrival, line.stopGroupSearchName, now))
            .filter((arrival) => arrival.lineNumber.trim().length > 0)
        );

        if (filtered.length === 0) continue;

        const next = filtered[0];
        const calls = await this.getTripCalls(next.trip.tripId, next.trip.startDate);
        const position = inferPosition(calls, now, next.stop.id);

        lineSummaries.push({
          configuredStopGroupName: line.stopGroupSearchName,
          configuredStopGroupAreaId: stopGroup.id,
          lineNumber: line.lineNumber,
          configuredOrigin: line.origin,
          configuredDestination: line.destination,
          nextArrivalTime: next.realtime,
          nextArrivalStationName: stopGroup.name,
          nextTripId: next.trip.tripId,
          nextTripStartDate: next.trip.startDate,
          position
        });
      }
    }

    if (lineSummaries.length === 0 && allArrivals.length === 0) {
      return null;
    }

    return {
      requestedAt: now,
      timeWindowMinutes: settings.timeWindowMinutes,
      arrivals: allArrivals.sort((a, b) => a.arrivalTime.getTime() - b.arrivalTime.getTime()),
      lines: lineSummaries,
      note: "SL bus GPS is not generally exposed via the public GTFS-RT VehiclePositions feed, so location is inferred from trip stop calls."
    };
  }

  async searchStations(query: string, limit: number = 15): Promise<string[]> {
    const trimmed = query.trim();
    if (!trimmed) return [];

    const json = await this.fetchJson(`stops/name/${encodeURIComponent(trimmed)}/?key=${encodeURIComponent(this.apiKey)}`) as {
      stop_groups?: Array<{ name?: string }>;
    };

    const names = (json.stop_groups ?? [])
      .map((group) => group.name ?? "")
      .filter((name) => name.trim().length > 0);

    return uniqueNonEmpty(names).slice(0, limit);
  }

  async getAllStations(forceRefresh: boolean = false): Promise<string[]> {
    if (!forceRefresh && SLApiClient.stationsMemoryCache && SLApiClient.stationsMemoryCache.length > 0) {
      return SLApiClient.stationsMemoryCache;
    }

    if (!forceRefresh) {
      const cached = this.readStationsCache();
      if (cached.length > 0) {
        SLApiClient.stationsMemoryCache = cached;
        return cached;
      }
    }

    let stations = await this.tryFetchAllStationsDirect();
    if (stations.length === 0) {
      stations = await this.fetchAllStationsBySearchSweep();
    }

    stations = uniqueNonEmpty(stations);
    if (stations.length > 0) {
      SLApiClient.stationsMemoryCache = stations;
      this.writeStationsCache(stations);
    }

    return stations;
  }

  async getRouteSuggestionsByStationName(stationName: string): Promise<StationRouteSuggestions> {
    const stopGroups = await this.findStopGroups(stationName);
    if (stopGroups.length === 0) {
      return { lineNumbers: [], origins: [], destinations: [], byLine: {} };
    }

    const buses: Arrival[] = [];
    for (const stopGroup of stopGroups) {
      const arrivals = await this.getArrivals(stopGroup.id);
      buses.push(...arrivals.filter((arrival) => equalIgnoreCase(arrival.route.transportMode, "BUS")));
    }

    const byLine: Record<string, { origins: string[]; destinations: string[]; pairs: Array<{ origin: string; destination: string }> }> = {};
    for (const bus of buses) {
      const line = bus.route.designation.trim();
      if (!line) continue;
      const origin = bus.route.originName.trim();
      const destination = bus.route.destinationName.trim();
      if (!byLine[line]) {
        byLine[line] = { origins: [], destinations: [], pairs: [] };
      }

      if (origin && !byLine[line].origins.some((v) => equalIgnoreCase(v, origin))) {
        byLine[line].origins.push(origin);
      }
      if (destination && !byLine[line].destinations.some((v) => equalIgnoreCase(v, destination))) {
        byLine[line].destinations.push(destination);
      }
      if (origin && destination && !byLine[line].pairs.some((p) => equalIgnoreCase(p.origin, origin) && equalIgnoreCase(p.destination, destination))) {
        byLine[line].pairs.push({ origin, destination });
      }
    }

    return {
      lineNumbers: uniqueNonEmpty([
        ...buses.map((arrival) => arrival.route.designation),
        ...Object.keys(byLine)
      ]),
      origins: uniqueNonEmpty(buses.map((arrival) => arrival.route.originName)),
      destinations: uniqueNonEmpty(buses.map((arrival) => arrival.route.destinationName)),
      byLine
    };
  }

  private async findStopGroups(search: string): Promise<StopGroup[]> {
    const json = await this.fetchJson(`stops/name/${encodeURIComponent(search)}/?key=${encodeURIComponent(this.apiKey)}`) as {
      stop_groups?: Array<{ id?: string; name?: string }>;
    };

    const groups = (json.stop_groups ?? [])
      .filter((g) => !!g.id && !!g.name)
      .map((g) => ({ id: g.id!, name: g.name! }));

    const exact = groups.filter((g) => equalIgnoreCase(g.name, search));
    if (exact.length > 0) return exact;

    return groups.filter((g) => containsIgnoreCase(g.name, search));
  }

  private async getArrivals(areaId: string): Promise<Arrival[]> {
    const json = await this.fetchJson(`arrivals/${encodeURIComponent(areaId)}?key=${encodeURIComponent(this.apiKey)}`) as {
      arrivals?: Array<Record<string, unknown>>;
    };

    return (json.arrivals ?? [])
      .map((raw) => this.parseArrival(raw))
      .filter((item): item is Arrival => item !== null);
  }

  private filterArrivals(arrivals: Arrival[], line: LineConfig, timeWindowMinutes: number, now: Date): Arrival[] {
    const from = new Date(now.getTime() - 60 * 1000);
    const to = new Date(now.getTime() + timeWindowMinutes * 60 * 1000);

    return arrivals
      .filter((a) => equalIgnoreCase(a.route.transportMode, "BUS"))
      .filter((a) => a.realtime >= from && a.realtime <= to)
      .filter((a) => equalIgnoreCase(a.route.designation, line.lineNumber))
      .filter((a) => containsIgnoreCase(a.route.originName, line.origin))
      .filter((a) => containsIgnoreCase(a.route.destinationName, line.destination))
      .sort((a, b) => a.realtime.getTime() - b.realtime.getTime());
  }

  private mapArrival(arrival: Arrival, configuredStopGroupName: string, now: Date): UpcomingBusArrival {
    const minutesUntilArrival = Math.round((arrival.realtime.getTime() - now.getTime()) / 60000);

    return {
      configuredStopGroupName,
      lineNumber: arrival.route.designation,
      origin: arrival.route.originName,
      destination: arrival.route.destinationName,
      direction: arrival.route.direction,
      arrivalTime: arrival.realtime,
      minutesUntilArrival,
      tripId: arrival.trip.tripId,
      tripStartDate: arrival.trip.startDate,
      stationName: arrival.stop.name,
      stationNumber: arrival.stop.id
    };
  }

  private async getTripCalls(tripId: string, startDate: string): Promise<TripCall[]> {
    const json = await this.fetchJson(`trips/${encodeURIComponent(tripId)}/${encodeURIComponent(startDate)}?key=${encodeURIComponent(this.apiKey)}`) as {
      calls?: Array<Record<string, unknown>>;
    };

    return (json.calls ?? []).map((call) => ({
      scheduledDeparture: asString(call.scheduledDeparture),
      realtimeDeparture: asString(call.realtimeDeparture),
      scheduledArrival: asString(call.scheduledArrival),
      realtimeArrival: asString(call.realtimeArrival),
      stopId: asString((call.stop as Record<string, unknown> | undefined)?.id),
      stopName: asString((call.stop as Record<string, unknown> | undefined)?.name)
    }));
  }

  private parseArrival(raw: Record<string, unknown>): Arrival | null {
    const realtime = parseDate(raw.realtime);
    if (!realtime) return null;

    const scheduled = parseDate(raw.scheduled) ?? realtime;
    const routeRaw = (raw.route ?? {}) as Record<string, unknown>;
    const tripRaw = (raw.trip ?? {}) as Record<string, unknown>;
    const stopRaw = (raw.stop ?? {}) as Record<string, unknown>;
    const originRaw = (routeRaw.origin ?? {}) as Record<string, unknown>;
    const destinationRaw = (routeRaw.destination ?? {}) as Record<string, unknown>;

    return {
      realtime,
      scheduled,
      route: {
        designation: asString(routeRaw.designation),
        transportMode: asString(routeRaw.transport_mode),
        direction: asString(routeRaw.direction),
        originName: asString(originRaw.name),
        destinationName: asString(destinationRaw.name)
      },
      trip: {
        tripId: asString(tripRaw.trip_id),
        startDate: asString(tripRaw.start_date)
      },
      stop: {
        id: asString(stopRaw.id),
        name: asString(stopRaw.name)
      }
    };
  }

  private async fetchJson(pathAndQuery: string): Promise<unknown> {
    const response = await fetch(`${this.baseUrl}/${pathAndQuery}`);
    if (!response.ok) {
      throw new Error(`SL API request failed: ${response.status} ${response.statusText}`);
    }

    const json = await response.json();
    if (isApiErrorPayload(json)) {
      throw new Error(`SL API error: ${json.errorCode}${json.errorDetail ? ` (${json.errorDetail})` : ""}`);
    }

    return json;
  }

  private async tryFetchAllStationsDirect(): Promise<string[]> {
    try {
      const json = await this.fetchJson(`stops?key=${encodeURIComponent(this.apiKey)}`) as {
        stop_groups?: Array<{ name?: string }>;
        stops?: Array<{ name?: string }>;
      };

      return uniqueNonEmpty([
        ...(json.stop_groups ?? []).map((x) => x.name ?? ""),
        ...(json.stops ?? []).map((x) => x.name ?? "")
      ]);
    } catch {
      return [];
    }
  }

  private async fetchAllStationsBySearchSweep(): Promise<string[]> {
    const buckets = [
      "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m",
      "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z",
      "0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
      "å", "ä", "ö"
    ];

    const all: string[] = [];
    for (const query of buckets) {
      try {
        const part = await this.searchStations(query, 200);
        all.push(...part);
      } catch {
        // Ignore partial errors and continue.
      }
    }
    return uniqueNonEmpty(all);
  }

  private readStationsCache(): string[] {
    try {
      const raw = localStorage.getItem(STATIONS_CACHE_KEY);
      if (!raw) return [];
      const parsed = JSON.parse(raw) as { savedAt?: number; stations?: string[] };
      if (!parsed?.savedAt || !Array.isArray(parsed.stations)) return [];
      if (Date.now() - parsed.savedAt > STATIONS_CACHE_TTL_MS) return [];
      return uniqueNonEmpty(parsed.stations);
    } catch {
      return [];
    }
  }

  private writeStationsCache(stations: string[]): void {
    try {
      localStorage.setItem(STATIONS_CACHE_KEY, JSON.stringify({
        savedAt: Date.now(),
        stations
      }));
    } catch {
      // Ignore storage failures.
    }
  }
}

function inferPosition(calls: TripCall[], now: Date, targetStopId: string): InferredBusPosition | undefined {
  const withTimes = calls
    .map((call, index) => ({ call, index, time: bestTime(call) }))
    .filter((item): item is { call: TripCall; index: number; time: Date } => item.time !== undefined);

  if (withTimes.length === 0) return undefined;

  const last = [...withTimes]
    .filter((item) => item.time.getTime() <= now.getTime())
    .sort((a, b) => a.time.getTime() - b.time.getTime())
    .at(-1);

  const next = [...withTimes]
    .filter((item) => item.time.getTime() > now.getTime())
    .sort((a, b) => a.time.getTime() - b.time.getTime())[0];

  let stopsRemainingToTarget: number | undefined;
  if (next) {
    const targetIndex = calls.findIndex((c) => equalIgnoreCase(c.stopId, targetStopId));
    if (targetIndex >= 0) {
      const currentIndex = last?.index ?? next.index;
      if (currentIndex <= targetIndex) {
        stopsRemainingToTarget = targetIndex - currentIndex;
      }
    }
  }

  return {
    lastStopName: last?.call.stopName ?? "(unknown)",
    nextStopName: next?.call.stopName ?? "(unknown)",
    stopsRemainingToTarget,
    targetStationNumber: targetStopId
  };
}

function bestTime(call: TripCall): Date | undefined {
  return parseDate(call.realtimeArrival)
    ?? parseDate(call.realtimeDeparture)
    ?? parseDate(call.scheduledArrival)
    ?? parseDate(call.scheduledDeparture);
}

function parseDate(value: unknown): Date | undefined {
  if (typeof value !== "string" || value.trim() === "") return undefined;
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? undefined : date;
}

function asString(value: unknown): string {
  return typeof value === "string" ? value : "";
}

function equalIgnoreCase(a: string | undefined, b: string | undefined): boolean {
  if (!a || !b) return false;
  return a.localeCompare(b, undefined, { sensitivity: "base" }) === 0;
}

function containsIgnoreCase(a: string | undefined, b: string | undefined): boolean {
  if (!a || !b) return false;
  return foldText(a).includes(foldText(b));
}

function foldText(value: string): string {
  return value
    .normalize("NFD")
    .replace(/\p{Diacritic}/gu, "")
    .toLocaleLowerCase();
}

function uniqueNonEmpty(values: string[]): string[] {
  const result: string[] = [];
  for (const value of values) {
    const trimmed = value.trim();
    if (!trimmed) continue;
    if (!result.some((item) => item.localeCompare(trimmed, undefined, { sensitivity: "accent" }) === 0)) {
      result.push(trimmed);
    }
  }
  return result;
}

function isApiErrorPayload(value: unknown): value is { errorCode: string; errorDetail?: string } {
  if (!value || typeof value !== "object") return false;
  const record = value as Record<string, unknown>;
  return typeof record.errorCode === "string";
}

File: SL.Lib/SLApiClient.cs
using System;
using System.Globalization;
using System.Net.Http;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace SL.Lib;

public class SLApiClient : ISLApiClient
{
    private readonly HttpClient Http;
    private readonly string ApiKey;
    public SLApiClient(HttpClient httpClient, string apiKey)
    {
        Http = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        ApiKey = apiKey ?? throw new ArgumentNullException(nameof(apiKey));
    }

    public async Task<List<StopGroup>> FindStopGroupsAsync(string search)
    {
        var url = $"stops/name/{Uri.EscapeDataString(search)}/?key={Uri.EscapeDataString(ApiKey)}";
        var json = await Http.GetStringAsync(url);

        using var doc = JsonDocument.Parse(json);
        if (!doc.RootElement.TryGetProperty("stop_groups", out var groupsEl) || groupsEl.ValueKind != JsonValueKind.Array)
            return new List<StopGroup>();

        List<StopGroup> result = new List<StopGroup>();

        foreach (var g in groupsEl.EnumerateArray())
        {
            var id = g.GetProperty("id").GetString();
            var name = g.GetProperty("name").GetString();
            result.Add(new StopGroup(id, name));
        }

        return result;
    }

    public async Task<List<Arrival>> GetArrivalsAsync(string areaId)
    {
        var url = $"arrivals/{Uri.EscapeDataString(areaId)}?key={Uri.EscapeDataString(ApiKey)}";
        var json = await Http.GetStringAsync(url);

        using var doc = JsonDocument.Parse(json);
        if (!doc.RootElement.TryGetProperty("arrivals", out var arrEl) || arrEl.ValueKind != JsonValueKind.Array)
            return new List<Arrival>();

        var list = new List<Arrival>();
        foreach (var a in arrEl.EnumerateArray())
        {
            var realtimeStr = a.GetProperty("realtime").GetString();
            var scheduledStr = a.GetProperty("scheduled").GetString();
            if (!DateTimeOffset.TryParse(realtimeStr, CultureInfo.InvariantCulture, DateTimeStyles.AssumeLocal, out var realtime))
                continue;
            DateTimeOffset.TryParse(scheduledStr, CultureInfo.InvariantCulture, DateTimeStyles.AssumeLocal, out var scheduled);

            var route = ParseRoute(a.GetProperty("route"));
            var trip = ParseTrip(a.GetProperty("trip"));
            var stop = ParseStop(a.GetProperty("stop"));

            list.Add(new Arrival(realtime, scheduled, route, trip, stop));
        }
        return list;
    }

    public async Task<TripDetailResponse?> GetTripDetailsAsync(string tripId, string startDate)
    {
        var url = $"trips/{Uri.EscapeDataString(tripId)}/{Uri.EscapeDataString(startDate)}?key={Uri.EscapeDataString(ApiKey)}";
        var json = await Http.GetStringAsync(url);

        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;
        if (root.ValueKind != JsonValueKind.Object) return null;

        var resp = new TripDetailResponse();

        if (root.TryGetProperty("timestamp", out var ts) && ts.ValueKind == JsonValueKind.String)
            resp.Timestamp = ts.GetString();

        if (root.TryGetProperty("route", out var routeEl) && routeEl.ValueKind == JsonValueKind.Object)
        {
            resp.Route = new RouteInfo
            {
                Designation = routeEl.TryGetProperty("designation", out var rdes) && rdes.ValueKind == JsonValueKind.String ? rdes.GetString() : null,
                TransportMode = routeEl.TryGetProperty("transport_mode", out var rtm) && rtm.ValueKind == JsonValueKind.String ? rtm.GetString() : null,
                TransportModeCode = routeEl.TryGetProperty("transport_mode_code", out var rtmc) && rtmc.ValueKind == JsonValueKind.Number ? rtmc.GetInt32() : null,
                Direction = routeEl.TryGetProperty("direction", out var rdir) && rdir.ValueKind == JsonValueKind.String ? rdir.GetString() : null,
                Origin = routeEl.TryGetProperty("origin", out var rorig) && rorig.ValueKind == JsonValueKind.Object ? new NamedStop(rorig.GetProperty("id").GetString(), rorig.GetProperty("name").GetString()) : null,
                Destination = routeEl.TryGetProperty("destination", out var rdst) && rdst.ValueKind == JsonValueKind.Object ? new NamedStop(rdst.GetProperty("id").GetString(), rdst.GetProperty("name").GetString()) : null
            };
        }

        if (root.TryGetProperty("trip", out var tripEl) && tripEl.ValueKind == JsonValueKind.Object)
        {
            resp.Trip = new TripInfo
            {
                TripId = tripEl.TryGetProperty("trip_id", out var tid) && tid.ValueKind == JsonValueKind.String ? tid.GetString() : null,
                StartDate = tripEl.TryGetProperty("start_date", out var sd) && sd.ValueKind == JsonValueKind.String ? sd.GetString() : null,
                TechnicalNumber = tripEl.TryGetProperty("technical_number", out var tn) && tn.ValueKind == JsonValueKind.Number ? tn.GetInt32() : null
            };
        }

        if (root.TryGetProperty("calls", out var callsEl) && callsEl.ValueKind == JsonValueKind.Array)
        {
            foreach (var ce in callsEl.EnumerateArray())
            {
                var call = new TripCall
                {
                    ScheduledDeparture = ce.TryGetProperty("scheduledDeparture", out var sd) && sd.ValueKind == JsonValueKind.String ? sd.GetString() : null,
                    RealtimeDeparture = ce.TryGetProperty("realtimeDeparture", out var rd) && rd.ValueKind == JsonValueKind.String ? rd.GetString() : null,
                    ScheduledArrival = ce.TryGetProperty("scheduledArrival", out var sa) && sa.ValueKind == JsonValueKind.String ? sa.GetString() : null,
                    RealtimeArrival = ce.TryGetProperty("realtimeArrival", out var ra) && ra.ValueKind == JsonValueKind.String ? ra.GetString() : null,
                };

                if (ce.TryGetProperty("stop", out var stopEl) && stopEl.ValueKind == JsonValueKind.Object)
                {
                    call.Stop = new TripStop
                    {
                        Id = stopEl.TryGetProperty("id", out var sid) && sid.ValueKind == JsonValueKind.String ? sid.GetString() : null,
                        AreaId = stopEl.TryGetProperty("area_id", out var aid) && aid.ValueKind == JsonValueKind.String ? aid.GetString() : null,
                        Name = stopEl.TryGetProperty("name", out var sname) && sname.ValueKind == JsonValueKind.String ? sname.GetString() : null,
                        Lat = stopEl.TryGetProperty("lat", out var slat) && slat.ValueKind == JsonValueKind.Number ? slat.GetDouble() : null,
                        Lon = stopEl.TryGetProperty("lon", out var slon) && slon.ValueKind == JsonValueKind.Number ? slon.GetDouble() : null
                    };
                }

                resp.Calls.Add(call);
            }
        }

        return resp;
    }

    public async Task<List<UpcomingBusQueryResult>> GetUpcomingBusesAsync(
        string stopGroupSearchName,
        TimeSpan timeWindow,
        IReadOnlyCollection<string>? origins = null,
        IReadOnlyCollection<string>? destinations = null,
        DateTimeOffset? now = null)
    {
        List<UpcomingBusQueryResult> result = new List<UpcomingBusQueryResult>();

        if (string.IsNullOrWhiteSpace(stopGroupSearchName))
            throw new ArgumentException("Stop group search name is required.", nameof(stopGroupSearchName));
        if (timeWindow <= TimeSpan.Zero)
            throw new ArgumentOutOfRangeException(nameof(timeWindow), "Time window must be greater than zero.");

        var current = now ?? DateTimeOffset.Now;
        var stopGroups = await FindStopGroupsAsync(stopGroupSearchName);
        if (stopGroups is null || stopGroups.Count == 0) return result;

        foreach(var stopGroup in stopGroups)        
        {
            var arrivals = await GetArrivalsAsync(stopGroup.Id);

            var filtered = arrivals
                .Where(a => string.Equals(a.Route.TransportMode, "BUS", StringComparison.OrdinalIgnoreCase))
                .Where(a => a.Realtime >= current.AddMinutes(-1))
                .Where(a => a.Realtime <= current.Add(timeWindow))
                .Where(a => origins is null || origins.Count == 0 || MatchAny(a.Route.Origin?.Name, origins))
                .Where(a => destinations is null || destinations.Count == 0 || MatchAny(a.Route.Destination?.Name, destinations))
                .OrderBy(a => a.Realtime)
                .ToList();

            var arrivalsResult = filtered.Select(a => new UpcomingBusArrival(
                LineNumber: a.Route.Designation ?? string.Empty,
                Origin: a.Route.Origin?.Name ?? string.Empty,
                Destination: a.Route.Destination?.Name ?? string.Empty,
                Direction: a.Route.Direction,
                ArrivalTime: a.Realtime,
                MinutesUntilArrival: (int)Math.Round((a.Realtime - current).TotalMinutes),
                TripId: a.Trip.TripId,
                TripStartDate: a.Trip.StartDate,
                StationName: a.Stop.Name,
                StationNumber: a.Stop.Id
            )).ToList();

            var lines = new List<UpcomingBusLineSummary>();
            foreach (var lineGroup in filtered.GroupBy(a => a.Route.Designation ?? string.Empty).OrderBy(g => g.Key))
            {
                var next = lineGroup.OrderBy(a => a.Realtime).First();
                InferredBusPosition? position = null;
                var tripDetails = await GetTripDetailsAsync(next.Trip.TripId, next.Trip.StartDate);
                if (tripDetails?.Calls is not null && tripDetails.Calls.Count > 0)
                {
                    var inferred = InferPosition(tripDetails.Calls, current, next.Stop.Id);
                    if (inferred is not null)
                    {
                        position = new InferredBusPosition(
                            LastStopName: inferred.LastStopName,
                            NextStopName: inferred.NextStopName,
                            StopsRemainingToTarget: inferred.StopsRemainingToTarget,
                            TargetStationNumber: next.Stop.Id
                        );
                    }
                }

                lines.Add(new UpcomingBusLineSummary(
                    LineNumber: lineGroup.Key,
                    NextArrivalTime: next.Realtime,
                    NextArrivalStationName: stopGroup.Name,
                    NextTripId: next.Trip.TripId,
                    NextTripStartDate: next.Trip.StartDate,
                    Position: position
                ));
            }

            result.Add(new UpcomingBusQueryResult(
                StopGroup: stopGroup,
                RequestedAt: current,
                TimeWindow: timeWindow,
                Arrivals: arrivalsResult,
                Lines: lines,
                Note: "SL bus GPS is not generally exposed via the public GTFS-RT VehiclePositions feed, so location is inferred from trip stop calls."
            ));
        }

        return result;
        
        static bool MatchAny(string? value, IReadOnlyCollection<string> patterns)
        {
            if (string.IsNullOrWhiteSpace(value)) return false;
            foreach (var pattern in patterns)
            {
                if (!string.IsNullOrWhiteSpace(pattern) &&
                    value.Contains(pattern, StringComparison.OrdinalIgnoreCase))
                    return true;
            }
            return false;
        }
    }

    // -------------------- Parsing helpers & models --------------------

    private Route ParseRoute(JsonElement r)
    {
        return new Route(
            Name: r.TryGetProperty("name", out var n) ? n.GetString() : null,
            Designation: r.TryGetProperty("designation", out var d) ? d.GetString() : null,
            TransportMode: r.TryGetProperty("transport_mode", out var tm) ? tm.GetString() : null,
            Direction: r.TryGetProperty("direction", out var dir) ? dir.GetString() : null,
            Origin: r.TryGetProperty("origin", out var o) ? new NamedStop(o.GetProperty("id").GetString(), o.GetProperty("name").GetString()) : null,
            Destination: r.TryGetProperty("destination", out var dst) ? new NamedStop(dst.GetProperty("id").GetString(), dst.GetProperty("name").GetString()) : null
        );
    }

    private Trip ParseTrip(JsonElement t)
    {
        return new Trip(
            TripId: t.GetProperty("trip_id").GetString() ?? "",
            StartDate: t.GetProperty("start_date").GetString() ?? ""
        );
    }

    private Stop ParseStop(JsonElement s)
    {
        return new Stop(
            Id: s.GetProperty("id").GetString() ?? "",
            Name: s.GetProperty("name").GetString() ?? ""
        );
    }

    private static InferredPosition? InferPosition(List<TripCall> calls, DateTimeOffset now, string targetStopId)
    {
        DateTimeOffset? BestTime(TripCall c)
        {
            if (TryParse(c.RealtimeArrival, out var ra)) return ra;
            if (TryParse(c.RealtimeDeparture, out var rd)) return rd;
            if (TryParse(c.ScheduledArrival, out var sa)) return sa;
            if (TryParse(c.ScheduledDeparture, out var sd)) return sd;
            return null;
        }

        var withTimes = calls
            .Select((c, idx) => new { Call = c, Index = idx, Time = BestTime(c) })
            .Where(x => x.Time is not null)
            .ToList();

        if (withTimes.Count == 0) return null;

        var last = withTimes.Where(x => x.Time!.Value <= now).OrderBy(x => x.Time).LastOrDefault();
        var next = withTimes.Where(x => x.Time!.Value > now).OrderBy(x => x.Time).FirstOrDefault();

        int? remaining = null;
        if (next is not null)
        {
            var targetIdx = calls.FindIndex(c => string.Equals(c.Stop?.Id, targetStopId, StringComparison.OrdinalIgnoreCase));
            if (targetIdx >= 0)
            {
                var currentIdx = last?.Index ?? next.Index;
                if (currentIdx <= targetIdx)
                    remaining = targetIdx - currentIdx;
            }
        }

        return new InferredPosition(
            LastStopName: last?.Call.Stop?.Name ?? "(unknown)",
            NextStopName: next?.Call.Stop?.Name ?? "(unknown)",
            StopsRemainingToTarget: remaining
        );

        static bool TryParse(string? s, out DateTimeOffset dto)
        {
            dto = default;
            return !string.IsNullOrWhiteSpace(s) &&
                   DateTimeOffset.TryParse(s, CultureInfo.InvariantCulture, DateTimeStyles.AssumeLocal, out dto);
        }
    }

    private record InferredPosition(string LastStopName, string NextStopName, int? StopsRemainingToTarget);

}

// Public model types exported by the library
public record StopGroup(string Id, string Name);

public record Arrival(
    DateTimeOffset Realtime,
    DateTimeOffset Scheduled,
    Route Route,
    Trip Trip,
    Stop Stop
);

public record Route(
    string? Name,
    string? Designation,
    string? TransportMode,
    string? Direction,
    NamedStop? Origin,
    NamedStop? Destination
);

public record Trip(string TripId, string StartDate);

public record Stop(string Id, string Name);

public record TripDetailResponse
{
    [JsonPropertyName("timestamp")]
    public string? Timestamp { get; set; }

    [JsonPropertyName("route")]
    public RouteInfo? Route { get; set; }

    [JsonPropertyName("trip")]
    public TripInfo? Trip { get; set; }

    [JsonPropertyName("calls")]
    public List<TripCall> Calls { get; set; } = new();
}

public record RouteInfo
{
    [JsonPropertyName("designation")]
    public string? Designation { get; set; }

    [JsonPropertyName("transport_mode")]
    public string? TransportMode { get; set; }

    [JsonPropertyName("transport_mode_code")]
    public int? TransportModeCode { get; set; }

    [JsonPropertyName("direction")]
    public string? Direction { get; set; }

    [JsonPropertyName("origin")]
    public NamedStop? Origin { get; set; }

    [JsonPropertyName("destination")]
    public NamedStop? Destination { get; set; }
}

public record TripInfo
{
    [JsonPropertyName("trip_id")]
    public string? TripId { get; set; }

    [JsonPropertyName("start_date")]
    public string? StartDate { get; set; }

    [JsonPropertyName("technical_number")]
    public int? TechnicalNumber { get; set; }
}

public record NamedStop(
    [property: JsonPropertyName("id")] string? Id,
    [property: JsonPropertyName("name")] string? Name
);

public record TripCall
{
    [JsonPropertyName("scheduledDeparture")]
    public string? ScheduledDeparture { get; set; }

    [JsonPropertyName("realtimeDeparture")]
    public string? RealtimeDeparture { get; set; }

    [JsonPropertyName("scheduledArrival")]
    public string? ScheduledArrival { get; set; }

    [JsonPropertyName("realtimeArrival")]
    public string? RealtimeArrival { get; set; }

    [JsonPropertyName("stop")]
    public TripStop? Stop { get; set; }
}

public record TripStop
{
    [JsonPropertyName("id")]
    public string? Id { get; set; }

    [JsonPropertyName("area_id")]
    public string? AreaId { get; set; }

    [JsonPropertyName("name")]
    public string? Name { get; set; }

    [JsonPropertyName("lat")]
    public double? Lat { get; set; }

    [JsonPropertyName("lon")]
    public double? Lon { get; set; }
}

public record UpcomingBusQueryResult(
    StopGroup StopGroup,
    DateTimeOffset RequestedAt,
    TimeSpan TimeWindow,
    List<UpcomingBusArrival> Arrivals,
    List<UpcomingBusLineSummary> Lines,
    string Note
);

public record UpcomingBusArrival(
    string LineNumber,
    string Origin,
    string Destination,
    string? Direction,
    DateTimeOffset ArrivalTime,
    int MinutesUntilArrival,
    string TripId,
    string TripStartDate,
    string StationName,
    string StationNumber
);

public record UpcomingBusLineSummary(
    string LineNumber,
    DateTimeOffset NextArrivalTime,
    string NextArrivalStationName,
    string NextTripId,
    string NextTripStartDate,
    InferredBusPosition? Position
);

public record InferredBusPosition(
    string LastStopName,
    string NextStopName,
    int? StopsRemainingToTarget,
    string TargetStationNumber
);
